// @linode/validation/lib/buckets.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UpdateBucketAccessSchema = exports.UploadCertificateSchema = exports.CreateBucketSchema = void 0;

var _yup = require("yup");

const CreateBucketSchema = (0, _yup.object)({
  label: (0, _yup.string)().required('Label is required.').matches(/^\S*$/, 'Label must not contain spaces.').ensure().min(3, 'Label must be between 3 and 63 characters.').max(63, 'Label must be between 3 and 63 characters.'),
  cluster: (0, _yup.string)().required('Cluster is required.')
});
exports.CreateBucketSchema = CreateBucketSchema;
const UploadCertificateSchema = (0, _yup.object)({
  certificate: (0, _yup.string)().required('Certificate is required.'),
  private_key: (0, _yup.string)().required('Private key is required.')
});
exports.UploadCertificateSchema = UploadCertificateSchema;
const UpdateBucketAccessSchema = (0, _yup.object)({
  acl: (0, _yup.string)().oneOf(['private', 'public-read', 'authenticated-read', 'public-read-write']).notRequired(),
  cors_enabled: (0, _yup.boolean)().notRequired()
});
exports.UpdateBucketAccessSchema = UpdateBucketAccessSchema;
