// @linode/validation/lib/databases.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.resetPasswordSchema = exports.updateDatabaseSchema = exports.createDatabaseSchema = exports.maintenanceScheduleSchema = void 0;

var _yup = require("yup");

const LABEL_MESSAGE = 'Label must be between 3 and 32 characters';
const maintenanceScheduleSchema = (0, _yup.object)({
  day: (0, _yup.mixed)() // .required('Day is required')
  .oneOf(['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']),
  window: (0, _yup.mixed)() // .required('Maintenance window is required')
  .oneOf(['W0', 'W2', 'W4', 'W6', 'W8', 'W10', 'W12', 'W14', 'W16', 'W18', 'W20', 'W22'])
}).notRequired().default(undefined);
exports.maintenanceScheduleSchema = maintenanceScheduleSchema;
const createDatabaseSchema = (0, _yup.object)({
  label: (0, _yup.string)().notRequired().min(3, LABEL_MESSAGE).max(32, LABEL_MESSAGE),
  region: (0, _yup.string)().required('Region is required'),
  type: (0, _yup.string)().required('Type is required'),
  root_password: (0, _yup.string)().required('Root password is required'),
  tags: (0, _yup.array)().of((0, _yup.string)()),
  maintenance_schedule: maintenanceScheduleSchema
});
exports.createDatabaseSchema = createDatabaseSchema;
const updateDatabaseSchema = (0, _yup.object)({
  label: (0, _yup.string)().notRequired().min(3, LABEL_MESSAGE).max(32, LABEL_MESSAGE),
  tags: (0, _yup.array)().of((0, _yup.string)()).notRequired(),
  maintenance_schedule: maintenanceScheduleSchema.notRequired()
});
exports.updateDatabaseSchema = updateDatabaseSchema;
const resetPasswordSchema = (0, _yup.object)({
  root_password: (0, _yup.string)().required('Root password is required')
});
exports.resetPasswordSchema = resetPasswordSchema;
