// @linode/validation/lib/images.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateImageSchema = exports.uploadImageSchema = exports.createImageSchema = exports.baseImageSchema = void 0;

var _yup = require("yup");

const labelSchema = (0, _yup.string)().max(50, 'Length must be 50 characters or less.').matches(/^[a-zA-Z0-9,.?\-_\s']+$/, 'Image labels cannot contain special characters.');
const baseImageSchema = (0, _yup.object)().shape({
  label: labelSchema.notRequired(),
  description: (0, _yup.string)().notRequired().min(1).max(65000)
});
exports.baseImageSchema = baseImageSchema;
const createImageSchema = baseImageSchema.shape({
  disk_id: (0, _yup.number)().typeError('Disk is required.').required('Disk is required.')
});
exports.createImageSchema = createImageSchema;
const uploadImageSchema = baseImageSchema.shape({
  label: labelSchema.required('Label is required.'),
  region: (0, _yup.string)().required('Region is required.')
});
exports.uploadImageSchema = uploadImageSchema;
const updateImageSchema = (0, _yup.object)().shape({
  label: (0, _yup.string)().notRequired().max(50, 'Length must be 50 characters or less.').matches(/^[a-zA-Z0-9,.?\-_\s']+$/, 'Image labels cannot contain special characters.'),
  description: (0, _yup.string)().notRequired().max(65000, 'Length must be 65000 characters or less.')
});
exports.updateImageSchema = updateImageSchema;
