// @linode/validation/lib/kubernetes.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createKubeClusterSchema = exports.clusterLabelSchema = exports.nodePoolSchema = void 0;

var _yup = require("yup");

const nodePoolSchema = (0, _yup.object)().shape({
  type: (0, _yup.string)(),
  count: (0, _yup.number)()
});
exports.nodePoolSchema = nodePoolSchema;
const clusterLabelSchema = (0, _yup.string)().required('Label is required.')
/**
 * This regex is adapted from the API docs. Kubernetes does
 * not allow underscores.
 */
.matches(/^[a-zA-Z0-9-]+$/, 'Cluster labels cannot contain special characters, spaces, or underscores.').min(3, 'Length must be between 3 and 32 characters.').max(32, 'Length must be between 3 and 32 characters.');
exports.clusterLabelSchema = clusterLabelSchema;
const createKubeClusterSchema = (0, _yup.object)().shape({
  label: clusterLabelSchema,
  region: (0, _yup.string)().required('Region is required.'),
  k8s_version: (0, _yup.string)().required('Kubernetes version is required.'),
  node_pools: (0, _yup.array)().of(nodePoolSchema).min(1, 'Please add at least one node pool.')
});
exports.createKubeClusterSchema = createKubeClusterSchema;
