// @linode/validation/lib/networking.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.shareAddressesSchema = exports.assignAddressesSchema = exports.allocateIPSchema = exports.updateIPSchema = void 0;

var _yup = require("yup");

const updateIPSchema = (0, _yup.object)().shape({
  rdns: (0, _yup.string)().notRequired().nullable(true)
});
exports.updateIPSchema = updateIPSchema;
const allocateIPSchema = (0, _yup.object)().shape({
  type: (0, _yup.string)().required().matches(/^ipv4$/, 'Only IPv4 address may be allocated through this endpoint.'),
  public: (0, _yup.boolean)().required(),
  linode_id: (0, _yup.number)().required()
});
exports.allocateIPSchema = allocateIPSchema;
const assignAddressesSchema = (0, _yup.object)().shape({
  region: (0, _yup.string)().required(),
  assignments: (0, _yup.array)().of((0, _yup.object)()).required()
});
exports.assignAddressesSchema = assignAddressesSchema;
const shareAddressesSchema = (0, _yup.object)().shape({
  linode_id: (0, _yup.number)().required(),
  ips: (0, _yup.array)().of((0, _yup.string)())
});
exports.shareAddressesSchema = shareAddressesSchema;
