// @linode/validation/lib/profile.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateProfileSchema = exports.createSSHKeySchema = exports.createPersonalAccessTokenSchema = void 0;

var _yup = require("yup");

const createPersonalAccessTokenSchema = (0, _yup.object)({
  scopes: (0, _yup.string)(),
  expiry: (0, _yup.string)(),
  label: (0, _yup.string)().min(1, 'Label must be between 1 and 100 characters.').max(100, 'Label must be between 1 and 100 characters.')
});
exports.createPersonalAccessTokenSchema = createPersonalAccessTokenSchema;
const createSSHKeySchema = (0, _yup.object)({
  label: (0, _yup.string)().required('Label is required.').min(1, 'Label must be between 1 and 64 characters.').max(64, 'Label must be between 1 and 64 characters.').trim(),
  ssh_key: (0, _yup.string)()
});
exports.createSSHKeySchema = createSSHKeySchema;
const updateProfileSchema = (0, _yup.object)({
  email: (0, _yup.string)().email(),
  timezone: (0, _yup.string)(),
  email_notifications: (0, _yup.boolean)(),
  authorized_keys: (0, _yup.array)().of((0, _yup.string)()),
  restricted: (0, _yup.boolean)(),
  two_factor_auth: (0, _yup.boolean)(),
  lish_auth_method: (0, _yup.string)().oneOf(['password_keys', 'keys_only', 'disabled']),
  authentication_type: (0, _yup.string)().oneOf(['password', 'github'])
});
exports.updateProfileSchema = updateProfileSchema;
