// @linode/validation/lib/records.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateRecordSchema = exports.createRecordSchema = void 0;

var _yup = require("yup");

const recordBaseSchema = (0, _yup.object)().shape({
  name: (0, _yup.string)().max(100, 'Record name must be 100 characters or less.'),
  target: (0, _yup.string)(),
  priority: (0, _yup.number)().min(0, 'Priority must be between 0 and 255.').max(255, 'Priority must be between 0 and 255.'),
  weight: (0, _yup.number)(),
  port: (0, _yup.number)(),
  service: (0, _yup.string)().nullable(true),
  protocol: (0, _yup.string)().nullable(true),
  ttl_sec: (0, _yup.number)(),
  tag: (0, _yup.string)()
});
const validRecordTypes = ['A', 'AAAA', 'NS', 'MX', 'CNAME', 'TXT', 'SRV', 'PTR', 'CAA'];
const createRecordSchema = recordBaseSchema.shape({
  type: (0, _yup.mixed)().required('Type is required.').oneOf(validRecordTypes)
});
exports.createRecordSchema = createRecordSchema;
const updateRecordSchema = recordBaseSchema.shape({});
exports.updateRecordSchema = updateRecordSchema;
