// @linode/validation/lib/stackscripts.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateStackScriptSchema = exports.stackScriptSchema = void 0;

var _yup = require("yup");

const stackScriptSchema = (0, _yup.object)({
  script: (0, _yup.string)().required('Script is required.'),
  label: (0, _yup.string)().required('Label is required.').min(3, 'Label must be between 3 and 128 characters.').max(128, 'Label must be between 3 and 128 characters.'),
  images: (0, _yup.array)().of((0, _yup.string)()).required('An image is required.'),
  description: (0, _yup.string)(),
  is_public: (0, _yup.boolean)(),
  rev_note: (0, _yup.string)()
});
exports.stackScriptSchema = stackScriptSchema;
const updateStackScriptSchema = (0, _yup.object)({
  script: (0, _yup.string)(),
  label: (0, _yup.string)().min(3, 'Label must be between 3 and 128 characters.').max(128, 'Label must be between 3 and 128 characters.'),
  images: (0, _yup.array)().of((0, _yup.string)()).min(1, 'An image is required.'),
  description: (0, _yup.string)(),
  is_public: (0, _yup.boolean)(),
  rev_note: (0, _yup.string)()
});
exports.updateStackScriptSchema = updateStackScriptSchema;
