// @linode/validation/lib/support.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createReplySchema = exports.createSupportTicketSchema = void 0;

var _yup = require("yup");

const createSupportTicketSchema = (0, _yup.object)({
  summary: (0, _yup.string)().required('Summary is required.').min(1, 'Summary must be between 1 and 64 characters.').max(64, 'Summary must be between 1 and 64 characters.').trim(),
  description: (0, _yup.string)().required('Description is required.').min(1, 'Description must be between 1 and 64,000 characters.').max(64000, 'Description must be between 1 and 64,000 characters.').trim(),
  domain_id: (0, _yup.number)(),
  linode_id: (0, _yup.number)(),
  longviewclient_id: (0, _yup.number)(),
  nodebalancer_id: (0, _yup.number)(),
  volume_id: (0, _yup.number)()
});
exports.createSupportTicketSchema = createSupportTicketSchema;
const createReplySchema = (0, _yup.object)({
  description: (0, _yup.string)().required('Description is required.').min(1, 'Description must be between 1 and 65,535 characters.').max(65535, 'Description must be between 1 and 65,535 characters.').trim()
});
exports.createReplySchema = createReplySchema;
