// @linode/validation/lib/volumes.schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AttachVolumeSchema = exports.UpdateVolumeSchema = exports.ResizeVolumeSchema = exports.CloneVolumeSchema = exports.CreateVolumeSchema = void 0;

var _yup = require("yup");

var _constants = require("./constants");

const createSizeValidation = (minSize = 10) => (0, _yup.number)().integer().typeError(`Size must be a number`).min(minSize, `Size must be between ${minSize} and ${_constants.MAX_VOLUME_SIZE}.`).max(_constants.MAX_VOLUME_SIZE, `Size must be between ${minSize} and ${_constants.MAX_VOLUME_SIZE}.`).required(`A size is required.`); // @todo this should be used in CreateVolumeForm and CreateVolumeFromLinodeForm
// export const tag = string()
//   .min(3, "Tags must be between 3 and 50 characters.")
//   .max(50, "Tags must be between 3 and 50 characters.")


const CreateVolumeSchema = (0, _yup.object)({
  region: (0, _yup.string)().when('linode_id', {
    is: id => id === undefined || id === '',
    then: (0, _yup.string)().required('Must provide a region or a Linode ID.')
  }),
  linode_id: (0, _yup.number)(),
  size: createSizeValidation(10),
  label: (0, _yup.string)().required('Label is required.').ensure().trim().min(1, 'Label must be between 1 and 32 characters.').max(32, 'Label must be 32 characters or less.'),
  config_id: (0, _yup.number)().typeError('Config ID must be a number.'),
  tags: (0, _yup.array)().of((0, _yup.string)())
});
exports.CreateVolumeSchema = CreateVolumeSchema;
const CloneVolumeSchema = (0, _yup.object)({
  label: (0, _yup.string)().required()
});
exports.CloneVolumeSchema = CloneVolumeSchema;

const ResizeVolumeSchema = (minSize = 10) => (0, _yup.object)({
  size: createSizeValidation(minSize)
});

exports.ResizeVolumeSchema = ResizeVolumeSchema;
const UpdateVolumeSchema = (0, _yup.object)({
  label: (0, _yup.string)().required()
});
exports.UpdateVolumeSchema = UpdateVolumeSchema;
const AttachVolumeSchema = (0, _yup.object)({
  linode_id: (0, _yup.number)().required(),
  config_id: (0, _yup.number)().required()
});
exports.AttachVolumeSchema = AttachVolumeSchema;
